### ISSUE TITLE
Update to major version 3 of mbedtls

### ISSUE THREAD

**Issue by suntorytimed** [lite-xl/lite-xl-plugin-manager#35]
I tried building with mbedtls 3.4.1, but mbedtls [deprecated and dropped net.h in 2021](https://github.com/Mbed-TLS/mbedtls/pull/4226) and replaced it by net_sockets.h.

This results in a compilation error on openSUSE with mbedtls 3.4.1

```
[   23s] ../src/lpm.c:36:10: fatal error: mbedtls/net.h: No such file or directory
[   23s]    36 | #include <mbedtls/net.h>
[   23s]       |          ^~~~~~~~~~~~~~~
[   23s] compilation terminated.
```

mbedtls 2 will reach EOL end of 2024.

**Comment by suntorytimed** [comment]
Changing the include to mbedtls/net_sockets.h results in following warnings and error:

```
[   13s] ../src/lpm.c: In function ‘lpm_hash’:
[   13s] ../src/lpm.c:109:3: warning: implicit declaration of function ‘mbedtls_sha256_starts_ret’; did you mean ‘mbedtls_sha256_starts’? [-Wimplicit-function-declaration]
[   13s]   109 |   mbedtls_sha256_starts_ret(&hash_ctx, 0);
[   13s]       |   ^~~~~~~~~~~~~~~~~~~~~~~~~
[   13s]       |   mbedtls_sha256_starts
[   13s] ../src/lpm.c:119:7: warning: implicit declaration of function ‘mbedtls_sha256_update_ret’; did you mean ‘mbedtls_sha256_update’? [-Wimplicit-function-declaration]
[   13s]   119 |       mbedtls_sha256_update_ret(&hash_ctx, chunk, bytes);
[   13s]       |       ^~~~~~~~~~~~~~~~~~~~~~~~~
[   13s]       |       mbedtls_sha256_update
[   13s] ../src/lpm.c:127:3: warning: implicit declaration of function ‘mbedtls_sha256_finish_ret’; did you mean ‘mbedtls_sha256_finish’? [-Wimplicit-function-declaration]
[   13s]   127 |   mbedtls_sha256_finish_ret(&hash_ctx, buffer);
[   13s]       |   ^~~~~~~~~~~~~~~~~~~~~~~~~
[   13s]       |   mbedtls_sha256_finish
[   13s] ../src/lpm.c: In function ‘lpm_socket_write’:
[   13s] ../src/lpm.c:794:35: warning: pointer targets in passing argument 2 of ‘mbedtls_ssl_write’ differ in signedness [-Wpointer-sign]
[   13s]   794 |     return mbedtls_ssl_write(ctx, buf, len);
[   13s]       |                                   ^~~
[   13s]       |                                   |
[   13s]       |                                   const char *
[   13s] In file included from ../src/lpm.c:34:
[   13s] /usr/include/mbedtls/ssl.h:4932:70: note: expected ‘const unsigned char *’ but argument is of type ‘const char *’
[   13s]  4932 | int mbedtls_ssl_write(mbedtls_ssl_context *ssl, const unsigned char *buf, size_t len);
[   13s]       |                                                 ~~~~~~~~~~~~~~~~~~~~~^~~
[   13s] ../src/lpm.c: In function ‘lpm_socket_read’:
[   13s] ../src/lpm.c:800:34: warning: pointer targets in passing argument 2 of ‘mbedtls_ssl_read’ differ in signedness [-Wpointer-sign]
[   13s]   800 |     return mbedtls_ssl_read(ctx, buf, len);
[   13s]       |                                  ^~~
[   13s]       |                                  |
[   13s]       |                                  char *
[   13s] /usr/include/mbedtls/ssl.h:4870:63: note: expected ‘unsigned char *’ but argument is of type ‘char *’
[   13s]  4870 | int mbedtls_ssl_read(mbedtls_ssl_context *ssl, unsigned char *buf, size_t len);
[   13s]       |                                                ~~~~~~~~~~~~~~~^~~
[   13s] ../src/lpm.c: In function ‘lpm_get’:
[   13s] ../src/lpm.c:835:8: warning: unused variable ‘response_code’ [-Wunused-variable]
[   13s]   835 |   long response_code;
[   13s]       |        ^~~~~~~~~~~~~
[   13s] ../src/lpm.c: At top level:
[   13s] ../src/lpm.c:413:28: warning: ‘ssl_context’ defined but not used [-Wunused-variable]
[   13s]   413 | static mbedtls_ssl_context ssl_context;
[   13s]       |                            ^~~~~~~~~~~
[   14s] [4/4] cc  -o lpm lpm.p/src_lpm.c.o -Wl,--as-needed -Wl,--no-undefined -Wl,--whole-archive -Wl,--start-group libmicrotar.a -Wl,--no-whole-archive -flto=auto -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -flto=auto /usr/lib64/libz.so -lmbedtls -lmbedx509 -lmbedcrypto /usr/lib64/libgit2.so /usr/lib64/libzip.so -lm -llua5.4 -Wl,--end-group
[   14s] FAILED: lpm 
[   14s] cc  -o lpm lpm.p/src_lpm.c.o -Wl,--as-needed -Wl,--no-undefined -Wl,--whole-archive -Wl,--start-group libmicrotar.a -Wl,--no-whole-archive -flto=auto -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -flto=auto /usr/lib64/libz.so -lmbedtls -lmbedx509 -lmbedcrypto /usr/lib64/libgit2.so /usr/lib64/libzip.so -lm -llua5.4 -Wl,--end-group
[   14s] /usr/lib64/gcc/x86_64-suse-linux/13/../../../../x86_64-suse-linux/bin/ld: /tmp/cc06s6IV.ltrans0.ltrans.o: in function `lpm_hash':
[   14s] <artificial>:(.text+0xa4): undefined reference to `mbedtls_sha256_starts_ret'
[   14s] /usr/lib64/gcc/x86_64-suse-linux/13/../../../../x86_64-suse-linux/bin/ld: <artificial>:(.text+0x104): undefined reference to `mbedtls_sha256_update_ret'
[   14s] /usr/lib64/gcc/x86_64-suse-linux/13/../../../../x86_64-suse-linux/bin/ld: <artificial>:(.text+0x12e): undefined reference to `mbedtls_sha256_update_ret'
[   14s] /usr/lib64/gcc/x86_64-suse-linux/13/../../../../x86_64-suse-linux/bin/ld: <artificial>:(.text+0x140): undefined reference to `mbedtls_sha256_finish_ret'
```

**Comment by adamharrison** [comment]
libgit2 does not support mbedtls3 at the moment, so before we upgrade, we have to wait for upstream.

**Comment by sbradnick** [comment]
Is it possible there were changes to mbedtls-2 v2.28.4 that would cause these issues for building `lpm`?

```
[   25s] [2/5] cc -Ilpm.p -I. -I.. -I../lib/microtar/src -I/usr/include/lua5.4 -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -fmessage-length=0 -grecord-gcc-switches -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -g -DOPENSSL_LOAD_CONF -DLPM_STATIC -MD -MQ lpm.p/src_lpm.c.o -MF lpm.p/src_lpm.c.o.d -o lpm.p/src_lpm.c.o -c ../src/lpm.c
[   25s] FAILED: lpm.p/src_lpm.c.o 
[   25s] cc -Ilpm.p -I. -I.. -I../lib/microtar/src -I/usr/include/lua5.4 -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -fmessage-length=0 -grecord-gcc-switches -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -g -DOPENSSL_LOAD_CONF -DLPM_STATIC -MD -MQ lpm.p/src_lpm.c.o -MF lpm.p/src_lpm.c.o.d -o lpm.p/src_lpm.c.o -c ../src/lpm.c
[   25s] ../src/lpm.c: In function ‘lpm_hash’:
[   25s] ../src/lpm.c:125:42: warning: pointer targets in passing argument 2 of ‘mbedtls_sha256_update_ret’ differ in signedness [-Wpointer-sign]
[   25s]      mbedtls_sha256_update_ret(&hash_ctx, data, len);
[   25s]                                           ^~~~
[   25s] In file included from ../src/lpm.c:30:0:
[   25s] /usr/include/mbedtls/sha256.h:122:5: note: expected ‘const unsigned char *’ but argument is of type ‘const char *’
[   25s]  int mbedtls_sha256_update_ret(mbedtls_sha256_context *ctx,
[   25s]      ^~~~~~~~~~~~~~~~~~~~~~~~~
[   25s] ../src/lpm.c: In function ‘lpm_fetch’:
[   25s] ../src/lpm.c:447:13: error: ‘git_fetch_options {aka struct <anonymous>}’ has no member named ‘depth’
[   25s]    fetch_opts.depth = 1;
[   25s]              ^
[   25s] ../src/lpm.c: In function ‘lpm_socket_write’:
[   25s] ../src/lpm.c:794:35: warning: pointer targets in passing argument 2 of ‘mbedtls_ssl_write’ differ in signedness [-Wpointer-sign]
[   25s]      return mbedtls_ssl_write(ctx, buf, len);
[   25s]                                    ^~~
[   25s] In file included from ../src/lpm.c:34:0:
[   25s] /usr/include/mbedtls/ssl.h:4183:5: note: expected ‘const unsigned char *’ but argument is of type ‘const char *’
[   25s]  int mbedtls_ssl_write(mbedtls_ssl_context *ssl, const unsigned char *buf, size_t len);
[   25s]      ^~~~~~~~~~~~~~~~~
[   25s] ../src/lpm.c: In function ‘lpm_socket_read’:
[   25s] ../src/lpm.c:800:34: warning: pointer targets in passing argument 2 of ‘mbedtls_ssl_read’ differ in signedness [-Wpointer-sign]
[   25s]      return mbedtls_ssl_read(ctx, buf, len);
[   25s]                                   ^~~
[   25s] In file included from ../src/lpm.c:34:0:
[   25s] /usr/include/mbedtls/ssl.h:4121:5: note: expected ‘unsigned char *’ but argument is of type ‘char *’
[   25s]  int mbedtls_ssl_read(mbedtls_ssl_context *ssl, unsigned char *buf, size_t len);
[   25s]      ^~~~~~~~~~~~~~~~
[   25s] ../src/lpm.c: In function ‘lpm_get’:
[   25s] ../src/lpm.c:835:8: warning: unused variable ‘response_code’ [-Wunused-variable]
[   25s]    long response_code;
[   25s]         ^~~~~~~~~~~~~
[   25s] At top level:
[   25s] ../src/lpm.c:413:28: warning: ‘ssl_context’ defined but not used [-Wunused-variable]
[   25s]  static mbedtls_ssl_context ssl_context;
[   25s]                             ^~~~~~~~~~~
```

**Comment by Guldoman** [comment]
> ```
> [   25s] ../src/lpm.c: In function ‘lpm_fetch’:
> [   25s] ../src/lpm.c:447:13: error: ‘git_fetch_options {aka struct <anonymous>}’ has no member named ‘depth’
> [   25s]    fetch_opts.depth = 1;
> [   25s]              ^
> ```

You might have an old version of libgit2.
AFAIK only the latest version supports this.

**Comment by sbradnick** [comment]
> You might have an old version of libgit2.

Ahh, ok - version of `libgit2` in play for Leap 15.5 (exhibiting this issue) is `1.3.0` ; in TW (doesn't have this issue) it's `1.7.0`

**Comment by adamharrison** [comment]
Yeah,  just use the submodules when building for now, should get you around this.

**Comment by sbradnick** [comment]
> Yeah, just use the submodules when building for now, should get you around this.

Not really an option for us when building packages for [OBS](https://build.opensuse.org); we can't do some type of `git clone https://github.com/lite-xl/lite-xl-plugin-manager.git --shallow-submodules --recurse-submodules` since the build environment doesn't have internet access during build time.

**Comment by adamharrison** [comment]
Hmm. Best option is probably to fork then, and replace the submodule with a copy directly.

**Comment by adamharrison** [comment]
Or we can put in that #if guard we talked about @Guldoman

**Comment by Guldoman** [comment]
I think they could specify `libgit2` as a source in the initial step and just download it before the build.
There might be concerns about possibly having multiple versions of `libgit2` on the same system, but if built statically it might be alright.

**Comment by sbradnick** [comment]
I'm not sure how familiar anyone here is with openSUSE Leap or Tumbleweed as well as SUSE Linux Enterprise [Server|Desktop], etc. but I'm not saying I need Lite-XL to provide `libgit2` in any way - just that Leap 15.5/15.6 (which is a bit behind "by design") seems to have `1.3.0` which doesn't seem to work with building `lpm`. But I did find that the [devel project](https://build.opensuse.org/package/show/devel:libraries:c_c++/libgit2) which hosts libgit2 has no issues building `1.7.1` for 15.5/15.6 ~ the issue is just that nothing post-`1.3.0` has been pushed forward for a few 15.X releases. All par for the course w/ how Leap is designed. And 99.9% of the reason I'm a TW user ;)

But since Lite-XL is offered in Leap, I'd like to get things updated to incorporate `lpm` vs. lite-xl[-colors,-plugins,-widgets] (my prior layout) as TW is now doing. I've got that all worked out in my home project, but it can be a long, winding road pushing things into Leap proper (whereas TW is much more fluid).
